### PR TITLE
Twitter verification & https refactoring

### LINKED ISSUE #38: Twitter Verification within TEE

### LINKED ISSUE #69: HTTPS support multi domain
So far, the https only support one domain after initialization, we need to support muliti domain and each domain has own handler. eg: twitter, discord

### PR DESCRIPTION
**Pull request by ziming-zung**
resolves #38 
resolves #69 

This PR:

- complete twitter verification (Encrypted tweets with tee shielding key is not completed, and now it is still returning to mock tweet). see at: https://github.com/litentry/tee-worker/blob/twitter-verification/core/https-request-handler/src/lib.rs#L154-L157
- refactor https request
- split `verify_identity` into 2 steps. see at: https://github.com/litentry/tee-worker/blob/twitter-verification/app-libs/stf/src/stf_sgx.rs#L467-L534

### CODE REVIEW

File: scripts/litentry/identity_test.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env bash
```
Comment (Kailai-Wang) [inline scripts/litentry/identity_test.sh:1]: maybe add this test to CI if it's ready?

File: cli/src/trusted_base_cli/commands/litentry/set_challenge_code.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
+/*
```
Comment (Kailai-Wang) [inline cli/src/trusted_base_cli/commands/litentry/set_challenge_code.rs]: Let's use litentry headers, same for other files

File: app-libs/stf/src/lib.rs [hunk 3]
```diff
@@ -291,8 +293,10 @@ impl TrustedCall {
 			TrustedCall::set_user_shielding_key(account, _, _) => account,
 			TrustedCall::link_identity(account, _, _, _, _) => account,
 			TrustedCall::unlink_identity(account, _, _) => account,
-			TrustedCall::verify_identity(account, _, _, _, _) => account,
+			TrustedCall::verify_identity_step1(account, _, _, _, _) => account,
+			TrustedCall::verify_identity_step2(account, _, _, _, _) => account,
 			TrustedCall::query_credit(account) => account,
+			TrustedCall::set_challenge_code(account, _, _, _) => account,
```
Comment (Kailai-Wang) [inline app-libs/stf/src/lib.rs:299]: I'm still thinking this should be merged into the `link_identity` TrustedCall handling, so when `link_identity` ends up with no errors, it should generate the challenge code (offchain) and call the extrinsics. What do you think?

btw I don't see where this `set_challenge_code` call is triggered in the current PR

File: app-libs/stf/src/stf_sgx_litentry.rs [hunk 4]
```diff
@@ -117,15 +120,63 @@ impl Stf {
...
 	}
+
+	pub fn verify_identity_step1(
+		_sender: AccountId,
+		target: AccountId,
+		identity: Identity,
+		validation_data: Web2ValidationData,
+		bn: ParentchainBlockNumber,
+	) -> StfResult<()> {
+		let code: Option<u32> = helpers::get_storage_double_map(
```
Comment (Kailai-Wang) [inline app-libs/stf/src/stf_sgx_litentry.rs]: Let's use pallet getters for direct access, these helper methods can be removed anytime soon, see https://github.com/integritee-network/worker/pull/967

File: app-libs/stf/src/stf_sgx.rs [hunk 5]
```diff
@@ -463,7 +464,34 @@ impl Stf {
...
+					identity,
+					validation_data,
+					bn,
+				) => {
+					// TODO support other validation_data
+					if let ValidationData::Web2(Web2ValidationData::Twitter(
+						TwitterValidationData { ref tweet_id },
+					)) = validation_data
+					{
+						Self::verify_identity_step1(
```
Comment (Kailai-Wang) [inline app-libs/stf/src/stf_sgx.rs]: A different fn name is better, something like `verify_identity_twitter`. In the future we can have a trait like `IdentityVerifier` and just write different implementations for different types of identities, that would be cleaner. 

Maybe we can create an issue for it.

File: core/https-client-daemon/src/lib.rs [hunk 6]
```diff
@@ -34,13 +34,17 @@ pub mod sgx_reexport_prelude {
...
-
+// pub mod https_client;
 pub use error::Result;
 
-use std::string::String;
+use codec::{Decode, Encode, MaxEncodedLen};
+use litentry_primitives::{Identity, Web2ValidationData};
 
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, MaxEncodedLen)]
 pub struct Request {
```
Comment (Kailai-Wang) [inline core/https-client-daemon/src/lib.rs:44]: Maybe call this `TwitterIdentityVerificationRequest` or any shorter name 😂  we'll have more requests that have different fields.

**Review by Kailai-Wang — COMMENTED** [review]
I haven't got time to look into the HTTP refactoring part, as I feel unwell today. Will follow up on this if I feel better tomorrow :(

File: scripts/litentry/identity_test.sh (hunk 1 above)
Comment (ziming-zung) [reply scripts/litentry/identity_test.sh:1]: There some problems in  this script, i will add it later.

File: app-libs/stf/src/lib.rs (hunk 3 above)
Comment (ziming-zung) [reply app-libs/stf/src/lib.rs:299]: I agree merging it into `link_identity` and `challenge_code` is generated by offchain. I think it's more reasonable.
Currently `set_challenge_code` call is only for testing.

File: core/https-client-daemon/src/lib.rs (hunk 6 above)
Comment (ziming-zung) [reply core/https-client-daemon/src/lib.rs:44]: Actually, I refactored the https further. 😂  I think this refactoring is what you think. 
See at:  https://github.com/litentry/tee-worker/commit/2d8c835c0d4ecb7386faeb8b7256413a2e308266#diff-131f0a349f74975ed71970e0a98d237e597f0130008bf5cf0c5c6e6943d81dad.

Maybe we can do this on the next PR. What do you think?

File: app-libs/stf/src/stf_sgx_litentry.rs (hunk 4 above)
Comment (ziming-zung) [reply app-libs/stf/src/stf_sgx_litentry.rs]: done

File: cli/src/trusted_base_cli/commands/litentry/set_challenge_code.rs (hunk 2 above)
Comment (ziming-zung) [reply cli/src/trusted_base_cli/commands/litentry/set_challenge_code.rs]: done

File: core/https-client-daemon/src/lib.rs (hunk 6 above)
Comment (Kailai-Wang) [reply core/https-client-daemon/src/lib.rs:44]: Let's do it in another PR. Btw there is a typo in the above link: `Web3IndentityVerification`

**Pull request merged**